File: referencia/python/fer.py
"""
Implementación de referencia de FER 1.1.0.

Convención de etiquetado para Facturas Electrónicas de retail (Panamá).
Especificación: spec/v1.1.0.md

Copyright 2026 Factura Fácil, S.A. — Apache-2.0
Sin dependencias externas. Compatible con Python 3.8+.
"""

from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass, field
from typing import Dict, Optional

VERSION_SPEC = "1.1.0"
VERSION_ESQUEMA_SOPORTADA = 1

CLAVES_DOCUMENTO = ("oc", "ref")
CLAVES_ITEM = ("cbar",)

# §3.3 — expresión regular canónica. El cierre debe coincidir en dígito.
_BLOQUE = re.compile(r"#FER(\d)#(.*?)#FER\1#", re.DOTALL)

# §7.1 — prefijo del perfil heredado. La insensibilidad a mayúsculas y tildes es una
# nota de implementación (§5.3.8), no un requisito de la convención.
_PREFIJO_HEREDADO = "orden de compra:"


@dataclass
class Lectura:
    """Resultado de leer un campo dInfEmFE."""

    perfil: str  # "fer" | "heredado"
    datos: Dict[str, str] = field(default_factory=dict)
    texto_libre: str = ""
    version_esquema: Optional[int] = None

    def __getitem__(self, clave: str) -> str:
        return self.datos[clave]

    def get(self, clave: str, defecto: Optional[str] = None) -> Optional[str]:
        return self.datos.get(clave, defecto)


def _sin_tildes(texto: str) -> str:
    descompuesto = unicodedata.normalize("NFD", texto)
    return "".join(c for c in descompuesto if unicodedata.category(c) != "Mn")


def _parsear_pares(contenido: str) -> Dict[str, str]:
    """§5.1.3 — divide el contenido del bloque en pares clave=valor."""
    datos: Dict[str, str] = {}
    for fragmento in contenido.split("|"):
        if "=" not in fragmento:
            continue
        clave, _, valor = fragmento.partition("=")  # divide en el primer "="
        clave = clave.strip().lower()
        valor = valor.strip()
        if not clave or not valor:
            continue  # §5.3.2 — valor vacío se trata como clave ausente
        if clave in datos:
            continue  # §5.3.1 — prevalece la primera ocurrencia
        datos[clave] = valor
    return datos


def _leer_heredado(valor: str, nivel: str) -> Lectura:
    """§7.1 — lectura del perfil heredado."""
    datos: Dict[str, str] = {}
    plano = valor.strip()

    if nivel == "documento":
        normalizado = _sin_tildes(plano).lower()
        pos = normalizado.find(_PREFIJO_HEREDADO)
        if pos != -1:
            resto = plano[pos + len(_PREFIJO_HEREDADO):].strip()
            partes = resto.split(None, 1)
            if partes:
                datos["oc"] = partes[0]
                if len(partes) > 1 and partes[1].strip():
                    datos["ref"] = partes[1].strip()
    elif nivel == "item":
        # §7.1 — el valor crudo es el contenido completo del campo. La convención no
        # define un criterio para distinguir un código de una nota: el formato
        # heredado no lo permite. Quien necesite esa distinción la resuelve con su
        # propio catálogo, fuera del alcance de FER.
        if plano:
            datos["cbar"] = plano

    return Lectura(perfil="heredado", datos=datos, texto_libre=plano)


def leer(valor: Optional[str], nivel: str = "documento") -> Lectura:
    """
    Lee un campo dInfEmFE conforme a FER 1.1.0.

    `valor` debe ser el contenido del nodo XML **ya des-escapado** por el parser
    (§5.1.1), no los bytes crudos del documento.

    `nivel` es "documento" (campo B29) o "item" (campo C19).

    Nunca lanza excepción por contenido malformado: descarta el bloque y cae al
    perfil heredado (§5.3.4). Esta implementación no trata un bloque ausente,
    sobrante o malformado como error (§5.3.5).
    """
    if nivel not in ("documento", "item"):
        raise ValueError("nivel debe ser 'documento' o 'item'")

    if not valor:
        return Lectura(perfil="heredado", datos={}, texto_libre="")

    coincidencia = _BLOQUE.search(valor)
    if coincidencia is None:
        return _leer_heredado(valor, nivel)

    datos = _parsear_pares(coincidencia.group(2))
    if not datos:
        # §5.3.4 — el bloque no produjo ningún par válido: se descarta.
        return _leer_heredado(valor, nivel)

    antes = valor[: coincidencia.start()].strip()
    despues = valor[coincidencia.end():].strip()
    texto_libre = " ".join(p for p in (antes, despues) if p)

    return Lectura(
        perfil="fer",
        datos=datos,
        texto_libre=texto_libre,
        version_esquema=int(coincidencia.group(1)),
    )


def escribir(nivel: str = "documento", texto_libre: str = "", **claves: str) -> str:
    """
    Construye el contenido de un campo dInfEmFE conforme a §4.

    >>> escribir(oc="4500000001", ref="TIENDA RETAIL X")
    '#FER1# oc=4500000001 | ref=TIENDA RETAIL X #FER1#'
    """
    centinela = f"#FER{VERSION_ESQUEMA_SOPORTADA}#"
    pares = []
    for clave, valor in claves.items():
        clave = clave.strip().lower()
        valor = "" if valor is None else str(valor).strip()
        if not valor:
            continue  # §5.3.2 — no se emiten claves con valor vacío
        if "|" in valor or re.search(r"#FER\d#", valor):
            raise ValueError(
                f"el valor de '{clave}' contiene un carácter reservado (§3.4)"
            )
        pares.append(f"{clave}={valor}")

    if not pares:
        return texto_libre.strip()

    bloque = f"{centinela} " + " | ".join(pares) + f" {centinela}"
    return f"{texto_libre.strip()} {bloque}".strip()

File: referencia/javascript/fer.js
/**
 * Implementación de referencia de FER 1.1.1.
 *
 * Convención de etiquetado para Facturas Electrónicas de retail (Panamá).
 * Especificación: spec/v1.1.1.md
 *
 * Copyright 2026 Factura Fácil, S.A. — Apache-2.0
 * Sin dependencias. Funciona en Node y en el navegador (ESM).
 */

export const VERSION_SPEC = "1.1.1";
export const VERSION_ESQUEMA_SOPORTADA = 1;

export const CLAVES_DOCUMENTO = ["oc", "ref"];
export const CLAVES_ITEM = ["cbar"];

// §3.3 — expresión regular canónica. El cierre debe coincidir en dígito.
const BLOQUE = /#FER(\d)#([\s\S]*?)#FER\1#/;

// §7.1 — prefijo del perfil heredado. La insensibilidad a mayúsculas y tildes es una
// nota de implementación (§5.3.8), no un requisito de la convención.
const PREFIJO_HEREDADO = "orden de compra:";

const RESERVADO_CENTINELA = /#FER\d#/;

function sinTildes(texto) {
  return texto.normalize("NFD").replace(/\p{Mn}/gu, "");
}

/**
 * §5.1.3 — divide el contenido del bloque en pares clave=valor.
 */
function parsearPares(contenido) {
  const datos = Object.create(null);
  for (const fragmento of contenido.split("|")) {
    const corte = fragmento.indexOf("="); // divide en el primer "="
    if (corte === -1) continue;
    const clave = fragmento.slice(0, corte).trim().toLowerCase();
    const valor = fragmento.slice(corte + 1).trim();
    if (!clave || !valor) continue; // §5.3.2 — valor vacío es clave ausente
    if (clave in datos) continue; // §5.3.1 — prevalece la primera ocurrencia
    datos[clave] = valor;
  }
  return datos;
}

/**
 * §7.1 — lectura del perfil heredado.
 */
function leerHeredado(valor, nivel) {
  const datos = Object.create(null);
  const plano = valor.trim();

  if (nivel === "documento") {
    const pos = sinTildes(plano).toLowerCase().indexOf(PREFIJO_HEREDADO);
    if (pos !== -1) {
      const resto = plano.slice(pos + PREFIJO_HEREDADO.length).trim();
      const partes = resto.split(/\s+/);
      if (partes.length && partes[0]) {
        datos.oc = partes[0];
        const ref = partes.slice(1).join(" ").trim();
        if (ref) datos.ref = ref;
      }
    }
  } else if (nivel === "item") {
    // §7.1 — el valor crudo es el contenido completo del campo. La convención no
    // define un criterio para distinguir un código de una nota: el formato heredado
    // no lo permite. Quien necesite esa distinción la resuelve con su propio
    // catálogo, fuera del alcance de FER.
    if (plano) datos.cbar = plano;
  }

  return { perfil: "heredado", datos, textoLibre: plano, versionEsquema: null };
}

/**
 * Lee un campo dInfEmFE conforme a FER 1.1.1.
 *
 * `valor` debe ser el contenido del nodo XML ya des-escapado por el parser (§5.1.1),
 * no los bytes crudos del documento.
 *
 * `nivel` es "documento" (campo B29) o "item" (campo C19).
 *
 * Nunca lanza por contenido malformado: descarta el bloque y cae al perfil heredado
 * (§5.3.4). Esta implementación no trata un bloque ausente, sobrante o malformado
 * como error (§5.3.5).
 *
 * @returns {{perfil: "fer"|"heredado", datos: Object, textoLibre: string, versionEsquema: number|null}}
 */
export function leer(valor, nivel = "documento") {
  if (nivel !== "documento" && nivel !== "item") {
    throw new Error('nivel debe ser "documento" o "item"');
  }

  if (!valor) {
    return {
      perfil: "heredado",
      datos: Object.create(null),
      textoLibre: "",
      versionEsquema: null,
    };
  }

  const m = BLOQUE.exec(valor);
  if (m === null) return leerHeredado(valor, nivel);

  const datos = parsearPares(m[2]);
  // §5.3.4 — el bloque no produjo ningún par válido: se descarta.
  if (Object.keys(datos).length === 0) return leerHeredado(valor, nivel);

  const antes = valor.slice(0, m.index).trim();
  const despues = valor.slice(m.index + m[0].length).trim();

  return {
    perfil: "fer",
    datos,
    textoLibre: [antes, despues].filter(Boolean).join(" "),
    versionEsquema: Number(m[1]),
  };
}

/**
 * Construye el contenido de un campo dInfEmFE conforme a §4.
 *
 *   escribir({ oc: "4500000001", ref: "TIENDA RETAIL X" })
 *   // '#FER1# oc=4500000001 | ref=TIENDA RETAIL X #FER1#'
 *
 * @param {Object} claves pares clave/valor a emitir
 * @param {string} textoLibre mensaje del emisor, opcional
 */
export function escribir(claves = {}, textoLibre = "") {
  const centinela = `#FER${VERSION_ESQUEMA_SOPORTADA}#`;
  const pares = [];

  for (const [claveCruda, valorCrudo] of Object.entries(claves)) {
    const clave = String(claveCruda).trim().toLowerCase();
    const valor = valorCrudo == null ? "" : String(valorCrudo).trim();
    if (!valor) continue; // §5.3.2 — no se emiten claves con valor vacío
    if (valor.includes("|") || RESERVADO_CENTINELA.test(valor)) {
      throw new Error(
        `el valor de "${clave}" contiene un carácter reservado (§3.4)`
      );
    }
    pares.push(`${clave}=${valor}`);
  }

  if (pares.length === 0) return textoLibre.trim();

  const bloque = `${centinela} ${pares.join(" | ")} ${centinela}`;
  return `${textoLibre.trim()} ${bloque}`.trim();
}
